lib/maths/analytics/CBoostedTreeHyperparameters.cc

```diff
@@ -94,14 +107,214 @@ void CBoostedTreeHyperparameters::resetSearch() {
     this->initializeTunableHyperparameters();
 }
 
-void CBoostedTreeHyperparameters::initializeSearch(const TAddInitialRangeFunc& addInitialRange) {
+void CBoostedTreeHyperparameters::initializeFineTuneSearchInterval(
+    const CInitializeFineTuneArguments& args,
+    TDoubleParameter& parameter) const {
+
+    if (parameter.valid(args.maxValue())) {
+
+        double maxValue{parameter.toSearchValue(args.maxValue())};
+        double minValue{maxValue - parameter.toSearchValue(args.searchInterval())};
+        double meanValue{(minValue + maxValue) / 2.0};
+        LOG_TRACE(<< "mean value = " << meanValue);
+
+        TVector3x1 fallback;
+        fallback(MIN_VALUE_INDEX) = minValue;
+        fallback(MID_VALUE_INDEX) = meanValue;
+        fallback(MAX_VALUE_INDEX) = maxValue;
+        auto interval = this->testLossLineSearch(args, minValue, maxValue).value_or(fallback);
+        args.truncateParameter()(interval);
+        LOG_TRACE(<< "search interval = [" << interval.toDelimited() << "]");
+
+        parameter.fixToRange(parameter.fromSearchValue(interval(MIN_VALUE_INDEX)),
+                             parameter.fromSearchValue(interval(MAX_VALUE_INDEX)));
+        parameter.set(parameter.fromSearchValue(interval(MID_VALUE_INDEX)));
+
+    } else {
+        parameter.fix();
+    }
+}
+
+CBoostedTreeHyperparameters::TOptionalVector3x1
+CBoostedTreeHyperparameters::testLossLineSearch(const CInitializeFineTuneArguments& args,
+                                                double intervalLeftEnd,
+                                                double intervalRightEnd) const {
+
+    // This has the following steps:
+    //   1. Coarse search the interval [intervalLeftEnd, intervalRightEnd] using
+    //      fixed steps,
+    //   2. Fine tune, via Bayesian Optimisation targeting expected improvement,
+    //      and stop if the expected improvement small compared to the current
+    //      minimum test loss,
+    //   3. Calculate the parameter interval which gives the lowest test losses,
+    //   4. Fit an OLS quadratic approximation to the test losses in the interval
+    //      from step 3 and use it to estimate the best parameter value,
+    //   5. Compare the size of the residual errors w.r.t. to the OLS curve from
+    //      step 4 with its variation over the interval from step 3 and truncate
+    //      the returned interval if we can determine there is a low chance of
+    //      missing the best solution by doing so.
```

Comment:
I think, it would improve readability if you either split this long function into individual steps or relate the individual steps in the comment to the places in the code where it happens.

Reply:
See [e5858c7](https://github.com/elastic/ml-cpp/pull/2125/commits/e5858c70df6ff8999fed13f67e7901d6665049f2).